### PR TITLE
Dont check on phase, but on pid. Register stacktrace as well (initial…

### PR DESCRIPTION
**Pull request by Tieske**
@thibaultcha  see code (untested). check on pid, not on phase, and add stacktraces.

### CODE REVIEW

**Comment by thibaultcha** [comment]
I consider this change actually harmful compared to the old implementation. The reason for only allowing the patched method to be called in `init_worker` is because:
1) it's not `init`
2) it's the earliest phase before workers start to process actual business logic

If called in `init` (that is, before the Lua VM is forked), then all workers will have the same seed, and this patch does nothing to prevent this scenario.
Secondly, this patch allows the method to be called in later phases (assuming it wasn't called before), but if that is the case, workers potentially have had time to generate some random numbers in their business logic that, well, weren't random, until this is called. 

tldr; sure, you won't be able to seed multiple times with this patch, but it does not enforce best practices for seeding in ngx_lua. The point of the current patch is to enforce best practises and make sure the workers are independently seeded, as early as possible, and only once.

**Comment by thibaultcha** [comment]
Prefer #1613 (also shorter, simpler btw).

**Comment by thibaultcha** [comment]
(Deleting the branch, keep a local copy if desired, reopen if debatable).

**Comment by Tieske** [comment]
> If called in `init` (that is, before the Lua VM is forked), then all workers will have the same seed, and this patch does nothing to prevent this scenario.

forking a worker changes the PID, hence generates a new seed. The original patch does not allow any random stuff to be done in the `init` phase (not just business logic, but also technical reasons for unicity are made impossible). 

> Secondly, this patch allows the method to be called in later phases (assuming it wasn't called before), but if that is the case, workers potentially have had time to generate some random numbers in their business logic that, well, weren't random, until this is called.

That is correct, good point. So maybe we shouldn't be patching randomseed, but random itself? Such that it seeds on the first call.

**Comment by thibaultcha** [comment]
Once forked, the PID sure changed but "hence generates a new seed", that, I don't see how? The method is not magically being called again, and the seed is cached at the module level, so all workers will have the same seed. 

It's ok, I suggest we keep the current method, we don't intend to seed in init so far, and one of today's PRs was suggesting to use OpenSSL, RAND_bytes, if you want to check it.
